adapter/serviceregistry/aggregate/controller.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aggregate
+
+import (
+	"github.com/golang/glog"
+
+	"istio.io/pilot/model"
+	"istio.io/pilot/platform"
+)
+
+// Registry specifies the collection of service registry related interfaces
+type Registry interface {
+	model.Controller
+	model.ServiceDiscovery
+	model.ServiceAccounts
+}
+
+// Controller aggregates data across different registries and monitors for changes
+type Controller struct {
+	registries map[platform.ServiceRegistry]Registry
+}
+
+// NewController creates a new Aggregate controller
+func NewController(registries map[platform.ServiceRegistry]Registry) *Controller {
+	return &Controller{
+		registries: registries,
+	}
+}
+
+// Services list declarations of all services in the system
+func (c *Controller) Services() []*model.Service {
+	services := make([]*model.Service, 0)
+	for _, r := range c.registries {
+		services = append(services, r.Services()...)
+	}
+	return services
+}
+
+// GetService retrieves a service by host name if it exists
+func (c *Controller) GetService(hostname string) (*model.Service, bool) {
+	for _, r := range c.registries {
+		if service, exists := r.GetService(hostname); exists {
+			return service, true
```

Comment:
so this simply assumes there is no overlap of service name in registry 1 or 2 or 3... etc.

Reply:
yep. which is the common case.